### ISSUE TITLE
FlatList blanks out when scrolled quickly (on large dataset)

### ISSUE THREAD

**Issue by mkashlev** [facebook/react-native#23442]
## 🐛 Bug Report
Contents of FlatList disappear upon quick scrolling

## To Reproduce
When a FlatList is scrolled with a pull to refresh and goes through multiple pages, and then scrolled up quickly, the content disappears.
Scroll down FlatList, going through multiple Pull to refresh instances until you have sufficiently large list. Then scroll up quickly, and the content disappears and does not come back. This behavior sometimes happens when you scroll down quickly as well.

## Expected Behavior
Content should not disappear when scrolled quickly up or down. This is the purpose of using FlatList over (deprecated) ListView. 

## Code Example
```
<FlatList
        style={styles.nudgeScrollView}
        onRefresh={onRefresh}
        refreshing={refreshing || false}
        onScroll={onScroll}
        onEndReached={onEndReached}
        onViewableItemsChanged={onViewableItemsChanged}
        scrollEventThrottle={400}
        data={nudges}
        keyExtractor={item => item.id}
        ListFooterComponent={footer}
        scrollEnabled={scrollEnabled}
        renderItem={({ item }) => (
          <NudgeCard
            nudge={item}
            {...{
              onOpenNudge,
              onArchiveNudge,
              allowListScroll,
              userId,
              archived,
            }}
          />
        )}
      />
    )
```

Even if onScroll or viewableItemsChanged props are commented out, the issue still remains.

## Environment

  React Native Environment Info:
    System:
      OS: macOS 10.14
      CPU: (8) x64 Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
      Memory: 24.15 MB / 16.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.11.0 - /usr/local/bin/node
      Yarn: 1.10.1 - /usr/local/bin/yarn
      npm: 6.4.1 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 12.1, macOS 10.14, tvOS 12.1, watchOS 5.1
      Android SDK:
        API Levels: 23, 25, 26, 27, 28
        Build Tools: 27.0.3, 28.0.2, 28.0.3
        System Images: android-19 | ARM EABI v7a, android-22 | Google APIs Intel x86 Atom, android-27 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: 3.2 AI-181.5540.7.32.5014246
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.6.3 => 16.6.3 
      react-native: 0.57.8 => 0.57.8 
    npmGlobalPackages:
      react-native-git-upgrade: 0.2.7

**Comment by mkashlev** [comment]
react native upgraded, but issue persists @react-native-bot

**Comment by lucsky** [comment]
@mkashlev From the FlatList documentation:
> In order to constrain memory and enable smooth scrolling, content is rendered asynchronously offscreen. This means it's possible to scroll faster than the fill rate and momentarily see blank content. This is a tradeoff that can be adjusted to suit the needs of each application, and we are working on improving it behind the scenes.

**Comment by mkashlev** [comment]
@lucsky Thanks for a quick response! This makes sense. One thing I noticed that in some cases, the screen does not get re-rendered, no matter how long I wait. Its a different thing from screen rendering being delayed (at least content shows up later). In my case the content does not show up.

**Comment by mkashlev** [comment]
Was wondering if anyone has experienced this area, or if I'm not doing it correctly? @sahrens

**Comment by akshetpandey** [comment]
@mkashlev : I have definitely faced the issue where the component never updates. Make sure your keyExtractor function is correct. Removing removeClippedSubviews and getItemLayout has fixed this for me but causes performance problems. Looks like there may be subtle bug in the flatlist implementation.

**Comment by xgAnd** [comment]
same error in this

**Comment by muhammadwfa** [comment]
I have the same issue, even for 100 items this happens to my flatlist

```
<FlatList
          ref={(ref) => (this.flatListRef = ref)}
          extraData={this.props.extraData}
          keyExtractor={this.keyExtractor}
          enableEmptySections
          automaticallyAdjustContentInsets={false}
          inverted={this.props.inverted}
          data={this.props.messages}
          style={styles.listStyle}
          renderItem={this.renderRow}
          onScroll={this.handleOnScroll}
          scrollEventThrottle={100}
        />
```

**Comment by muhammadwfa** [comment]
@mkashlev did you find any solutions for this issue?

**Comment by mkashlev** [comment]
@muhammadwfa - still no solution.

**Comment by Frekansapp** [comment]
I have this issue too and onreachend event calling with delay

**Comment by alloy** [comment]
As noted by @lucsky, the documentation mentions this trade-off and thus it is currently considered to work as intended. What the docs don’t do a great job at is explaining to you how you can fine-tune this behaviour to make the right trade-offs for your situation, something that I’m sure PRs for would be much appreciated.

In an effort to reduce the noise in the issue tracker, I’ll close this issue, but will leave you with a few links that were helpful to me recently:

* https://medium.com/sanjagh/how-to-optimize-your-react-native-flatlist-946490c8c49b
* https://github.com/filipemerker/flatlist-performance-tips

**Comment by johinsDev** [comment]
> @muhammadwfa - still no solution.

do you have some solution for this problem?

**Comment by muhammadwfa** [comment]
Now I am using scrollview, it works well.

**Comment by nihp** [comment]
Any solution for this in Flatlist

**Comment by Rainsho** [comment]
Keep watching.

**Comment by Rahulkishanm** [comment]
getItemLayout has fixed this for me.

**Comment by nihp** [comment]
@Rahulkishanm Can you give any code and how you have calculate the getItemLayout for entire list?

**Comment by nihp** [comment]
@Rahulkishanm getItemLayout props not worked for me 

 ```
 getItemLayout(data, index) {
  return (
    {
      length: deviceHeight,
      offset: (deviceHeight) * index,
      index
    }
  );
}

  renderGrid(item) {
    return(
      <TouchableOpacity activeOpacity = { .5 } >
        <View style={{backgroundColor: 'white', alignItems: 'center'}}>
          <FastImage style={{width: deviceWidth, height: deviceHeight, margin:6}}
            resizeMode={FastImage.resizeMode.contain}
            source={item.uri}/>
        </View>
      </TouchableOpacity>
    );
  }
```

**Comment by Shanie1331** [comment]
@nihp Can you please post a full code of your flat list implementation as It can be improved with some props that I can help you with.

**Comment by nihp** [comment]
@Shanie1331 I have updated my flatlist implementation above. Can you look into this?

**Comment by nihp** [comment]
Shanie1331 Any update

**Comment by Shanie1331** [comment]
Can you wait for a while as I need to replicate your implementation?

**Comment by nihp** [comment]
Okay.

Thanks

**Comment by tootsweet** [comment]
Hello,

We are experiencing this bug as well with RN61.

**Comment by vishnusurabhirao** [comment]
Me tooo

**Comment by SelvaraniAlagumuthu** [comment]
Hi, 
I am also faced while scrolling the list i got blank pages. After adding scrollview in flatlist now its working fine. Here my code please check may be its helpful. 
 <ScrollView>
            <View style={{ flex: 1}}> 
                <FlatList
                    data={data }
                    renderItem={({ item }) => <this.Item item={item} />}
                    keyExtractor={item => item.id}                   
                />
            </View>
            </ScrollView>
Thanks

**Comment by nihp** [comment]
@SelvaraniAlagumuthu Can you give your render Item method in brief

**Comment by androidsdkproject** [comment]
> Hi,
> I am also faced while scrolling the list i got blank pages. After adding scrollview in flatlist now its working fine. Here my code please check may be its helpful.
> 
> <View style={{ flex: 1}}>
> <FlatList
> data={data }
> renderItem={({ item }) => <this.Item item={item} />}
> keyExtractor={item => item.id}
> />
> 
> 
> Thanks

Can you share your code of renderItem , i am facing same issue.

**Comment by nihp** [comment]
Here I don't want to add scroll view for the screen. Any solution other
than this

**Comment by vishnusurabhirao** [comment]
<GridView
              showsHorizontalScrollIndicator={false}
              showsVerticalScrollIndicator={false}
              itemDimension={INV_GRID_CELL_WIDTH}
              spacing={INV_GRID_SPACING}
              items={items}
              style={{ flex: 1 }}
              itemContainerStyle={{ alignItems: "center", marginTop: INV_GRID_SPACING}}
              renderItem={item => (
                <View style={styles.sty}>
                  <View style={{ width: '100%' }}>
                    <TouchableOpacity style={{ width: '100%' }} onPress={}>
                      {method}
                      {method}
                    </TouchableOpacity>
                   {method}
                  </View>
			{method}
                        {method}
                        {method}
                </View>
              )}
            />

**Comment by androidsdkproject** [comment]
Thanks for replying.

*Thanks and Regards*
*Kamal Verma*
*+91-9509371638*

**Comment by NikGanyponko** [comment]
> I have this issue too and onreachend event calling with delay

Yeah! Thank you, I had a problem => in some cases flatList's callback onEndReached didn't call by scrolling to the end. So, I've added setTimeout on endReached and it works perfect.
